py/runai_model_streamer/runai_model_streamer/distributed_streamer/distributed_streamer.py

```diff
@@ -0,0 +1,502 @@
+from typing import List, Iterator, Optional
+from typing import Iterator, Optional, List
+import torch
+import torch.distributed as dist
+import glob
+import os
+from datetime import timedelta
+import socket
+import re
+
+from runai_model_streamer.file_streamer import (
+    FileStreamer,
+    FileChunks,
+)
+
+from runai_model_streamer.distributed_streamer.partition import (
+    partition,
+    get_total_number_of_chunks,
+    get_partition_policy,
+    log_partition_info,
+)
+
+from runai_model_streamer.s3_utils.s3_utils import (
+    S3Credentials,
+    is_s3_path,
+    is_gs_path,
+)
+
+import humanize
+from timeit import default_timer as timer
+
+DEFAULT_BROADCAST_TIMEOUT = timedelta(seconds=600)
+
+class DistributedStreamer:
+    def __init__(self) -> None:
+        self.file_streamer = FileStreamer()
+        self.device_str = None
+        self.is_distributed = False
+        self.partitions = {} # partitions of all the input file_stream_requests
+        self.rank_file_chunks_list = {} # post Partitioning FileChunks to be streamed by this rank
+        self.rank_dicts_map = {} # maps post partitioning FileChunks.id and chunk index to the original FileChunks.id and chunk index and chunk size
+        self.total_chunks_to_read = 0
+        self.rank = 0
+        self.original_group_rank = 0
+        self.distribution_group = None
+        self.reading_from_storage = False
+        self.local_group_global_ranks = []
+        self.is_error = False
+        self.my_global_rank = 0
+        self.groups_by_ranks = []
+        self.num_processes_on_node = None
+
+    def __enter__(self) -> DistributedStreamer:
+        self.file_streamer.__enter__()
+        return self
+
+    def __exit__(self, exc_type: any, exc_value: any, traceback: any) -> None:
+        if self.distribution_group and not self.is_error:
+            torch.distributed.barrier(group=self.distribution_group)
+            torch.distributed.destroy_process_group(group=self.distribution_group)
+            del self.distribution_group
+            self.distribution_group = None
+        return self.file_streamer.__exit__(exc_type, exc_value, traceback)
+
+    def get_group_size(self) -> int:
+        if not dist.is_initialized():
+            return 1
+        return dist.get_world_size()
+
+    def set_is_distributed(self, path: str, device: Optional[str] = None) -> None:
+        # check if distributed streaming should be used
+
+        # by default, use distributed streaming only for object storage paths
+        if path is not None:
+            self.is_distributed = is_s3_path(path) or is_gs_path(path)
+
+        # environment variable to override default distributed streaming
+        if os.environ.get("RUNAI_STREAMER_DIST") == "0":
+            self.is_distributed = False
+        elif os.environ.get("RUNAI_STREAMER_DIST") == "1":
+            self.is_distributed = True
+
+        # check if torch distributed is initialized and there are more than one process
```

Comment:
Lets have a warning log in case the user specifically requested dist but does not have a group